### PR TITLE
Address Sanitizer support

### PR DESCRIPTION
**Pull request by rayslava**
Address Sanitizer technology support is added by this pull request.
Current implementation of Address Sanitizer is capable of checking heap memory with current toolchain and heap and stack memory if patch for [shadow offset change](https://gcc.gnu.org/viewcvs/gcc?view=revision&revision=216773) is applied to GCC.

Currently it's tested on `qemu_16m` target since I did not receive appropriate device yet. The implementation will be updated if needed after on-device test is performed.

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/asan_test/asan_test_main.c:56: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/mm/asan/asan.c [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+  while (n-- > 0) {
+    *p++ = c;
+  }
+  return s;
+}
+
+__attribute__((no_sanitize_address))
+void asan_clean_shadow(void)
+{
+  unsigned char *addr = (unsigned char *) asan_mem_to_shadow((void*)CONFIG_RAM_START);
```
Comment (jeongchanKim) [inline os/mm/asan/asan.c]: Add space like below
(void *)

File: os/mm/asan/asan.c [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+{
+	asan_poison_shadow(address, size, ASAN_MALLOC_FREE);
+}
+__attribute__((no_sanitize_address))
+void asan_unpoison_shadow(const void *address, size_t size)
+{
+	asan_poison_shadow(address, size, 0);
+
+	if (size & ASAN_SHADOW_MASK) {
+		u8 *shadow = (u8 *) asan_mem_to_shadow(address + size);
```
Comment (jeongchanKim) [inline os/mm/asan/asan.c]: remove space after type casting

File: os/mm/asan/asan.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ ****************************************************************************/
+
+#ifndef __MM_ASAN_H
+#define __MM_ASAN_H
+#include <tinyara/config.h>
+#include <stdbool.h>
+
+#ifdef CONFIG_MM_ASAN_RT
+
+#define ASAN_SHADOW_OFFSET         (unsigned long) CONFIG_MM_ASAN_SHADOW_OFFSET
```
Comment (jeongchanKim) [inline os/mm/asan/asan.h]: remove space after type casting

File: os/mm/asan/asan.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#define ASAN_SHADOW_SCALE_SHIFT    3
+#define ASAN_SHADOW_SCALE_SIZE     (1UL << ASAN_SHADOW_SCALE_SHIFT)
+#define ASAN_SHADOW_MASK           (ASAN_SHADOW_SCALE_SIZE - 1)
+#define _RET_IP_                   (unsigned long)__builtin_return_address(0)
+#define __alias(symbol)            __attribute__((alias(#symbol)))
+#define BITS_PER_LONG              sizeof(long) * 8
+#define __round_mask(x, y)         ((__typeof__(x))((y)-1))
+#define round_up(x, y)             ((((x)-1) | __round_mask(x, y))+1)
```
Comment (jeongchanKim) [inline os/mm/asan/asan.h]: add space between operator and operand

File: os/mm/asan/Make.defs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#
+###########################################################################
+
+# Address Sanitizer run-time
+
+ifeq ($(CONFIG_MM_ASAN_RT),y)
+CSRCS += asan.c report.c
+CFLAGS+=-I$(TOPDIR)/kernel/
+
+# Add the kasan directory to the build
```
Comment (jeongchanKim) [inline os/mm/asan/Make.defs]: this directory is for asan, isn't it?

File: apps/examples/asan_test/asan_test_main.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	printf("a[2]@%08p: %x\n", &a[2], (int)a[2]);
+	printf("a[3]@%08p: %x\n", &a[3], (int)a[3]);
+	printf("b[3]@%08p: %x\n", &b[3], (int)b[3]);
+	printf("c[3]@%08p: %x\n", &c[3], (int)c[3]);
+	free(b);
+	printf("======================================================\n\n");
+	y = rand();
+	for (x = 0; x < 1; ++x) {
+	  y = rand() % 4096;
+	  int *d = (int*)realloc(a, y * sizeof(int));
```
Comment (jeongchanKim) [inline apps/examples/asan_test/asan_test_main.c]: Add a space like below
(int *)

File: os/mm/asan/asan.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+/*
+ * All functions below always inlined so compiler could
+ * perform better optimizations in each of __asan_loadX/__assn_storeX
+ * depending on memory access size X.
+ */
+__attribute__((no_sanitize_address))
+static inline bool memory_is_poisoned_1(unsigned long addr)
+{
+	s8 shadow_value = *(s8 *) asan_mem_to_shadow((void *)addr);
```
Comment (jeongchanKim) [inline os/mm/asan/asan.c]: check the spaces

File: os/mm/mm_heap/mm_initialize.c [hunk 8]
```diff
@@ -62,6 +62,10 @@
 
 #include <tinyara/mm/mm.h>
 
+#ifdef CONFIG_MM_ASAN_RT
+#include <asan/asan.h>
```
Comment (sunghan-chang) [inline os/mm/mm_heap/mm_initialize.c]: Is the asan.h os/mm/asan/asan.h? If it is, let's use " " instead of < >.

File: os/mm/mm_heap/mm_malloc.c [hunk 9]
```diff
@@ -64,6 +64,10 @@
 #ifdef CONFIG_DEBUG_MM_HEAPINFO
 #include  <tinyara/sched.h>
 #endif
+
+#ifdef CONFIG_MM_ASAN_RT
+#include <asan/asan.h>
```
Comment (sunghan-chang) [inline os/mm/mm_heap/mm_malloc.c]: same as previous

File: apps/shell/tash_command.c [hunk 10]
```diff
@@ -337,6 +337,9 @@ void tash_cmdlist_install(const tash_cmdlist_t list[])
 
 void tash_register_basic_cmds(void)
 {
+#ifdef CONFIG_MM_ASAN_RT
+  asan_unpoison_shadow(&tash_cmds_info, sizeof(struct tash_cmd_info_s));
```
Comment (sunghan-chang) [inline apps/shell/tash_command.c]: please check Tizen coding rule

File: os/kernel/clock/clock_systimespec.c [hunk 11]
```diff
@@ -186,7 +186,8 @@ int clock_systimespec(FAR struct timespec *ts)
 		/* Get the time since power-on in seconds and milliseconds */
 
 		msecs = TICK2MSEC(clock_systimer());
-		secs  = msecs / MSEC_PER_SEC;
+				secs  = msecs / MSEC_PER_SEC;
```
Comment (sunghan-chang) [inline os/kernel/clock/clock_systimespec.c]: Let's revert these two line changes.

File: os/kernel/sched/sched_releasetcb.c [hunk 12]
```diff
@@ -136,6 +136,9 @@ int sched_releasetcb(FAR struct tcb_s *tcb, uint8_t ttype)
 	int ret = OK;
 
 	if (tcb) {
+	  void * stack_ptr = tcb->stack_alloc_ptr;
```
Comment (sunghan-chang) [inline os/kernel/sched/sched_releasetcb.c]: please check indentation

File: os/kernel/task/task_setup.c [hunk 13]
```diff
@@ -594,6 +599,10 @@ static inline int task_stackargsetup(FAR struct task_tcb_s *tcb, FAR char *const
 
 		stackargv[i + 1] = str;
 		nbytes = strlen(argv[i]) + 1;
+#ifdef CONFIG_MM_ASAN_RT
+	mlldbg("Unpoisoning str %p, size %u\n", str, nbytes);
```
Comment (sunghan-chang) [inline os/kernel/task/task_setup.c]: alignment

**Comment by sunghan-chang** [comment]
I have a question. I saw only Samsung copyright and license at new added files. Is that Samsung implementation? not getting open source project and modifying?

**Comment by rayslava** [comment]
> Tizen coding rule

Where can I find them? There is no such file in `docs`, and I executed `os/tools/formatter.sh` script before commit.
> I saw only Samsung copyright and license at new added files. Is that Samsung implementation?

Yes, I created test files form scratch, are there any issues with that?

File: os/arch/arm/src/armv7-m/up_hardfault.c [hunk 14]
```diff
@@ -108,7 +108,6 @@
  *   exceptions that are performed in bad contexts.
  *
  ****************************************************************************/
-
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-m/up_hardfault.c]: Let's revert this change

**Comment by sunghan-chang** [comment]
@rayslava Please find the coding rule link on our [wiki](https://github.com/Samsung/TizenRT/wiki/Documentations#coding-rules).
And I mean new files on mm folder which is core of sanitizer. If it comes from open source project, it can be a violation of opensource project.

**Comment by seinfra** [comment]
Code Rule Check Result:
os/mm/asan/asan.c:320: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/mm/asan/asan.c:327: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Code Rule Check Result:
os/mm/asan/asan.c:323: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by dkson95** [comment]
LGTM :+1:

**Comment by rayslava** [comment]
I updated `asan_test`, freed memory poisoning procedure and fixed possible bug found by ASan.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/mm/asan/asan.c [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+	return s;
+}
+
+__attribute__((no_sanitize_address))
+void asan_clean_shadow(void)
+{
+	unsigned char *addr = (unsigned char *)asan_mem_to_shadow((void *)CONFIG_RAM_START);
+	int n = CONFIG_RAM_SIZE >> ASAN_SHADOW_SCALE_SHIFT;
+	while (n > 0) {
+		*(uint32_t *) addr = 0;
```
Comment (jeongchanKim) [inline os/mm/asan/asan.c]: remove space after type casting

File: apps/examples/asan_test/asan_test_main.c [hunk 16]
```diff
@@ -80,9 +48,18 @@ int asan_test_main(int argc, char *argv[])
 	printf("a[0]@%08p: %x\n", &a[0], (int)a[1]);
 	printf("a[2]@%08p: %x\n", &a[2], (int)a[2]);
 	printf("a[3]@%08p: %x\n", &a[3], (int)a[3]);
-	printf("b[3]@%08p: %x\n", &b[3], (int)b[3]);
+	printf("b[3]@%08p: %x\n", &b[4], (int)b[4]);
```
Comment (dkson95) [inline apps/examples/asan_test/asan_test_main.c]: -> printf("b[4]...

File: os/kernel/task/task_create.c [hunk 17]
```diff
@@ -128,6 +129,11 @@ static int thread_create(FAR const char *name, uint8_t ttype, int priority, int
 	pid_t pid;
 	int errcode;
 	int ret;
+	int tcb_size = sizeof(struct task_tcb_s);
+
+#ifndef CONFIG_DISABLE_PTHREAD
+	tcb_size = MAX(sizeof(struct task_tcb_s), sizeof(struct pthread_tcb_s));
```
[the remark this answers is not in the document]
Comment (rayslava) [reply os/kernel/task/task_create.c]: Please refer the commit message for this change.

I do understand that this function is not related to threads, but current implementation may produce very hard-to-catch bugs, and since the result might be used as `pthread_tcb_s` we have to ensure that there will be enough space for it.

File: os/mm/Kconfig [hunk 18]
```diff
@@ -166,3 +166,52 @@ config MM_SHM
...
+       bool "Build Address Sanitizer Run-Time library"
+       default y
+       ---help---
+		Build in support for Address Sanitizer run-time support
+		library.
+
+config MM_ASAN_SHADOW_OFFSET
+       hex "Shadow Memory Offset"
+       depends on MM_ASAN_RT
+       default 0x20100000
```
[the remark this answers is not in the document]
Comment (rayslava) [reply os/mm/Kconfig]: Technically there should be an addressable RAM space which should not be used by allocator (e.g. the end of RAM) with the size of 1/8 of total system RAM (currently ASan uses 8:1 scale factor).

Should I add this information into help message for the option?

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by rayslava** [comment]
The code of checking memory was taken from Linux kernel and therefore is GPLv2, do we need to rewrite it?

**Comment by juitem** [comment]
We can apply on dev(internal), not on public. We'll let you know the discussion results

File: os/kernel/task/task_create.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/kernel/task/task_create.c]: I don't understand what you want to say. As you agreed, this functions always uses task_tcb_s, never uses pthread_tcb_s. Why do you allocate maximun size between task and pthread?
If pthread_tcb_s is used inside that function, problem which we should fix is at another location. This is not a method of solving what you worried.

**Comment by dkson95** [comment]
> We can apply on dev(internal), not on public. 

Agreed on applying dev(internally), for now.

File: os/mm/mm_heap/mm_malloc.c [hunk 19]
```diff
@@ -106,6 +106,7 @@
  *  8-byte alignment of the allocated data is assured.
  *
  ****************************************************************************/
+__attribute__((no_sanitize_address))
```
Comment (girishks2000) [inline os/mm/mm_heap/mm_malloc.c]: This code makes it compiler dependent, might not work for non-gcc

**Comment by rayslava** [comment]
Okay, discarding this pull-request and applying internally.

File: os/mm/asan/report.c [hunk 20]
```diff
@@ -9,8 +9,9 @@
 #define SHADOW_BYTES_PER_BLOCK 1
 #define SHADOW_BLOCKS_PER_ROW 16
 #define SHADOW_BYTES_PER_ROW (SHADOW_BLOCKS_PER_ROW * SHADOW_BYTES_PER_BLOCK)
-#define SHADOW_ROWS_AROUND_ADDR 2
+#define SHADOW_ROWS_AROUND_ADDR 4
```
Comment (girishks2000) [inline os/mm/asan/report.c]: Lot of code with __atribite__ compiler directives below make it gnu dependent. It cannot be a generic solution
